csharp/src/Ice/InputStream.cs

```diff
@@ -642,23 +768,25 @@ public object ReadSerializable()
 
         /// <summary>Reads an empty encapsulation from the provided byte buffer.</summary>
         /// <param name="communicator">The communicator.</param>
+        /// <param name="encoding">The encoding of the buffer.</param>
         /// <param name="buffer">The byte buffer.</param>
-        internal static void ReadEmptyEncapsulation(Communicator communicator, ArraySegment<byte> buffer)
+        internal static void ReadEmptyEncapsulation(
+            Communicator communicator, Encoding encoding, ArraySegment<byte> buffer)
         {
-            var istr = new InputStream(communicator, buffer, startEncaps: true, 0);
+            var istr = new InputStream(communicator, encoding, buffer, startEncaps: true, 0);
             istr.SkipTaggedParams();
             istr.CheckEndOfBuffer();
         }
 
         /// <summary>Reads the contents of an encapsulation from the provided byte buffer.</summary>
         /// <param name="communicator">The communicator.</param>
+        /// <param name="encoding">The encoding of the buffer.</param>
         /// <param name="buffer">The byte buffer.</param>
         /// <param name="payloadReader">The reader used to read the payload of this encapsulation.</param>
-        internal static T ReadEncapsulation<T>(Communicator communicator,
-                                               ArraySegment<byte> buffer,
-                                               InputStreamReader<T> payloadReader)
+        internal static T ReadEncapsulation<T>(
+            Communicator communicator, Encoding encoding, ArraySegment<byte> buffer, InputStreamReader<T> payloadReader)
         {
-            var istr = new InputStream(communicator, buffer, startEncaps: true, 0);
+            var istr = new InputStream(communicator, encoding, buffer, startEncaps: true, 0);
```

Comment:
If we change this static method to ReadEncapsulationPayload we might be able to also get rid of the startEncaps parameter for the InputStream constructor. The check for whether or not the encoding is supported could directly be done here.